### PR TITLE
TE: don't compute wgrad when weights are frozen

### PR DESCRIPTION
**Pull request by kshitij12345**
Problem - Currently, TE executor just assumes that we will be computing gradients for both input and weight. This is true for full finetuning, but in PEFT scenario the weights are frozen i.e. requires_grad=False. This leads to poor performance as the TE executor ends up doing unnecessary work of computing gradients for weights.

Solution - Based on the `requires_grad` on TensorProxy for weight and bias, enable/disable the relevant gradient computation.

NOTE - Thunder doesn't reliable propagate the requires_grad on intermediate values (see https://github.com/Lightning-AI/lightning-thunder/issues/1768) so we just assume that requires_grad on `inp` is True. However, it should be safe to check the `requires_grad` on Parameters as they are inputs to the jitted function.

### CODE REVIEW

File: thunder/distributed/prims.py [hunk 1]
```diff
@@ -184,7 +184,14 @@ def synchronize_meta(
...
             unsharded_shape = a.shape[0] * group.size(), *a.shape[1:]
-            return TensorProxy(shape=unsharded_shape, like=a, distparallel_type=DistParallelType.REPLICATED)
+            # NOTE: We explicitly pass the `a.requires_grad` as we want to propogate the requires grad correctly on `nn.Parameter`'s
+            # as it maybe queried later.
+            return TensorProxy(
+                shape=unsharded_shape,
+                like=a,
+                distparallel_type=DistParallelType.REPLICATED,
+                requires_grad=a.requires_grad,
+            )
```
Comment (kshitij12345) [inline thunder/distributed/prims.py:194]: This change was required to get the TE FSDP tests working again.

Without this, the `requires_grad` on the sharded weight tensor is not propagated correctly.

**Comment by kshitij12345** [comment]
Ping @t-vi for review and stamp
